### PR TITLE
AWQ QuantizationMixin + SequentialPipeline

### LINKED ISSUE #1369: OOM (host) when running AWQ
**Describe the bug**
Running out of host memory when calibrating using AWQModifier.
The model I use is llama-3-instruct. Host memory is 126G.
Is this a exsisting problem that calibrating will consume too much memory? Thanks!

```
(py310_new) zjnyly@ric-MS-7D25:~/LLMs$ python awq_one_shot.py
INFO 04-22 23:14:37 __init__.py:190] Automatically detected platform cuda.
2025-04-22:23:14:38,762 INFO     [modeling.py:957] We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.85it/s]
Repo card metadata block was not found. Setting CardData to empty.
2025-04-22:23:14:42,634 WARNING  [repocard.py:108] Repo card metadata block was not found. Setting CardData to empty.
2025-04-22T23:14:44.900211+0800 | reset | INFO - Compression lifecycle reset
Logging all LLM Compressor modifier-level logs to sparse_logs/22-04-2025_23.14.44.log
2025-04-22:23:14:44,900 INFO     [logger.py:391] Logging all LLM Compressor modifier-level logs to sparse_logs/22-04-2025_23.14.44.log
2025-04-22T23:14:44.900749+0800 | from_modifiers | INFO - Creating recipe from modifiers
2025-04-22T23:14:46.201364+0800 | _calibrate | INFO - Running AWQModifier calibration with 256 samples...
 56%|█████████████████████████████████████████████▏                                   | 143/256 [00:53<00:43,  2.62it/s] 
Killed
```

**Environment**
Include all relevant environment information:
1. OS ubuntu 18.04
2. Python version 3.10.16
3. LLM Compressor version 0.5.1.dev26+g76af7af2 
4. compressed-tensors  0.9.4a20250414

### LINKED ISSUE #1409: [AWQ] Insane memory requirement: over 900GB for 32B model
I tried to quantized GLM-4-0414-32B: https://huggingface.co/THUDM/GLM-4-32B-0414

Recipe:
```python
recipe = [
    AWQModifier(
        bits=4,
        symmetric=False,
        # Read input->output from https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/models/glm4/modeling_glm4.py
        # which are somewhat easier than vllm ones as it's all in a single file
        mappings=[
            AWQMapping("re:.*input_layernorm", ["re:.*q_proj", "re:.*k_proj", "re:.*v_proj"]),
            AWQMapping("re:.*v_proj", ["re:.*o_proj"]),
            AWQMapping("re:.*post_attention_layernorm", ["re:.*gate_up_proj"]),
            AWQMapping("re:.*gate_up_proj", ["re:.*down_proj"]),
        ]
    ),
    QuantizationModifier(
        ignore=ignore_layers,
        config_groups={
            "group_0": QuantizationScheme(
                targets=["Linear"],
                weights=QuantizationArgs(
                    num_bits=4,
                    type=QuantizationType.INT,
                    dynamic=False,
                    symmetric=False,
                    strategy=QuantizationStrategy.GROUP,
                    group_size=128,
                ),
            ),
        },
    )
]
```

I tried using 128 samples as suggested in those slides ("Calibration set"): https://minjiazhang.github.io/courses/fall24-resource/slides/awq.pdf

However every sample memory usage grew by 1~5 GB leading in the end to over 900GB before I decided to give up on AWQ. Even with a swapfile, the time was spent in kernel swap in/out and IO and compute was slow to then frustratingly crash on Cuda OOM once that CPU part was solved.

Screenshot:
Image

 Side-note: couldn't the calibration be made multi-threaded?

### PR DESCRIPTION
**Pull request by brian-dellabetta**
SUMMARY:
- Add QuantizationMixin to AWQModifier so we don't have redundant inputs (num_bits, symmetric, group_size)
- Move AWQModifier to sequential pipelines, to avoid huge memory requirements of caching all activations at once.

Regression test results are acceptable, results are all roughly the same, and within stderr, see test plan below.

Resolves #1409 
Resolves #1369 
Related to #1383
Related to #1406 
Related to #1368 
Related to #1410 

More improvements split into #1435

TEST PLAN:
- [x] Rerun tests to validate
No regression in tests, comparing against those reported in [original AWQ PR](https://github.com/vllm-project/llm-compressor/pull/1177#issuecomment-2773113040). All gsm8k results are within stderr:

| Type            | gsm8k       | wikitext
| ------          | ------      | ----- 
| Old AWQ+QuantModifier Sym          | .1054, .1069     | 9.1931 
| New AWQ+QuantMixin Sym        | .1077, .1084 | 9.1841
| Old AWQ+QuantModifier Asym             | .1274, .1281 | 9.0281
| New AWQ+QuantMixin Asym        | .1312, .1350 | 9.0288

### CODE REVIEW

File: src/llmcompressor/modifiers/awq/base.py [hunk 1]
```diff
@@ -96,53 +112,138 @@ class AWQModifier(Modifier):
...
+
+        elif event.type_ == EventType.SEQUENTIAL_EPOCH_END:
+            self._concat_collected_activations()
+
+            with calibration_forward_context(state.model), HooksMixin.disable_hooks():
+                self._apply_smoothing(state.model)
+
+            self._activations.clear()
+
+        elif event.type_ == EventType.CALIBRATION_EPOCH_END:
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:222]: In order to support the basic pipeline, do `self._apply_smoothing(state.model)`

File: src/llmcompressor/modifiers/awq/base.py [hunk 2]
```diff
@@ -96,53 +112,138 @@ class AWQModifier(Modifier):
...
+        # This choice does not seem to have a meaningful impact on accuracy
+        state.model.apply(disable_quantization)
+        self._setup_activation_cache_hooks()
+
+    def on_event(self, state: State, event: Event, **kwargs):
+        if event.type_ == EventType.CALIBRATION_EPOCH_START:
+            if not self.started_:
+                self.on_start(state, None)
+
+        elif event.type_ == EventType.SEQUENTIAL_EPOCH_START:
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:211]: How we typically avoid having to have an epoch start event is by deleting accumulations as they're used

```python3
def sequential_epoch_end
   for activation in achivations:
       ... use activation
       del self._activations[...]
```

File: src/llmcompressor/modifiers/awq/base.py [hunk 3]
```diff
@@ -220,64 +324,47 @@ def _set_resolved_mappings(self, model: Module) -> None:
...
         """
 
-        def create_hook_fn(layer_name):
-            def hook_fn(module, inp, out):
-                inp = inp[0].cpu().detach()
-
-                if layer_name in self._scales:
-                    self._scales[layer_name].append(inp)
+        def create_cache_activation_hook(layer_name):
+            def cache_activation_hook_fn(
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:334]: nit: You can just define this hook as a class method, rather than creating a new instance of the function every time

File: src/llmcompressor/modifiers/utils/pytorch_helpers.py [hunk 4]
```diff
@@ -34,54 +28,6 @@ def apply_pad_mask_to_batch(batch: Dict[str, torch.Tensor]) -> Dict[str, torch.T
     return batch
 
 
-def run_calibration_forward(
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/utils/pytorch_helpers.py:37]: sweet

File: src/llmcompressor/modifiers/awq/base.py [hunk 5]
```diff
@@ -96,53 +112,138 @@ class AWQModifier(Modifier):
...
         """
 
+        # apply config to model and prepare calibration hooks
+        if QuantizationMixin.has_config(self):
+            QuantizationMixin.initialize_quantization(self, state.model)
+
         self._set_resolved_mappings(state.model)
 
         with calibration_forward_context(state.model):
             self._set_module_kwargs(state.model, state.data.calib)
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:191]: Can we get rid of this?

File: src/llmcompressor/modifiers/awq/base.py [hunk 6]
```diff
@@ -96,53 +112,138 @@ class AWQModifier(Modifier):
...
+            if not self.started_:
+                self.on_start(state, None)
+
+        elif event.type_ == EventType.SEQUENTIAL_EPOCH_START:
+            self._activations.clear()
+
+        elif event.type_ == EventType.SEQUENTIAL_EPOCH_END:
+            self._concat_collected_activations()
+
+            with calibration_forward_context(state.model), HooksMixin.disable_hooks():
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:217]: Consider moving this calibration_forward_context to within _apply_smoothing so that it only covers the area that you need it to

File: src/llmcompressor/modifiers/awq/base.py [hunk 7]
```diff
@@ -96,53 +112,138 @@ class AWQModifier(Modifier):
...
+        elif event.type_ == EventType.CALIBRATION_EPOCH_END:
+            if not self.ended_:
+                self.on_end(state, None)
+
+    def on_end(self, state: State, event: Event, **kwargs):
+        """
+        Finish calibrating by removing observers and calibration hooks
+        """
+        self.ended_ = True
+
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:231]: I typically add a warning here to check if there are any `_activations` that haven't been used to scale yet

File: src/llmcompressor/modifiers/awq/base.py (hunk 1 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:222]: thanks, added

File: src/llmcompressor/modifiers/awq/base.py (hunk 2 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:211]: updated to delete activations as they are used, and raise error if unused activations exist

File: src/llmcompressor/modifiers/awq/base.py [hunk 8]
```diff
@@ -220,64 +324,47 @@ def _set_resolved_mappings(self, model: Module) -> None:
...
+                layer, create_cache_activation_hook(mapping.smooth_name), "forward"
             )
-
-        run_calibration_forward(
-            model,
-            calibration_dataloader,
-        )
+            self._activation_hooks.add(hook)
 
     def _concat_collected_activations(self) -> None:
```
[the remark this answers is not in the document]
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:359]: removed

File: src/llmcompressor/modifiers/awq/base.py (hunk 5 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:191]: I'm hoping so, but we need it to run forward passes during smoothing. Bit different than GPTQ, AWQ needs the full output as well as the quantized output for the loss function. Maybe we can just cache output activations as well?

File: src/llmcompressor/modifiers/awq/base.py (hunk 7 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:231]: added

File: src/llmcompressor/modifiers/awq/base.py (hunk 6 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:214]: I think a better solution is dropping forward passes from smoothing entirely, by caching output activations. but let's chat about that

File: src/llmcompressor/modifiers/awq/base.py (hunk 3 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:337]: Is there a performance hit to doing that? If not, I prefer this as more scoped

Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:339]: Also note the `layer_name` input, i think we'll have to create a function every time because of this

File: src/llmcompressor/modifiers/awq/base.py [hunk 9]
```diff
@@ -96,53 +116,147 @@ class AWQModifier(Modifier):
...
+    _activations: Dict[str, List[torch.Tensor]] = PrivateAttr(default_factory=dict)
+    _activation_hooks: Set[RemovableHandle] = PrivateAttr(default_factory=set)
+    _module_kwargs: Dict = PrivateAttr(default_factory=dict)
+
+    @model_validator(mode="after")
+    def validate_model_after(model: "AWQModifier") -> "AWQModifier":
+        if not model.config_groups and not model.scheme:
+            raise ValueError("AWQ requires either a config_groups or a scheme")
+
+        # TODO better way to do this?
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:141]: You can use `config = self.resolve_config()` and then validate that

Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:141]: thanks!

File: src/llmcompressor/modifiers/awq/base.py (hunk 3 above)
Comment (kylesayrs) [reply src/llmcompressor/modifiers/awq/base.py:347]: Ah yeah, in other implementations I curry this using `partial` but this is good too

File: src/llmcompressor/modifiers/awq/base.py (hunk 9 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:141]: hmm, got a link? I don't see any `resolve_config` defined anywhere in the code

File: src/llmcompressor/modifiers/awq/base.py (hunk 6 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:225]: I think we'll need to keep the forward passes for now, so I moved this down into a lower scope in _apply_smoothing and _compute_loss

File: src/llmcompressor/modifiers/awq/base.py (hunk 3 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:354]: 🍛

File: src/llmcompressor/modifiers/awq/base.py (hunk 9 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: found `.resolve_quantization_config`, thanks!

File: src/llmcompressor/modifiers/awq/base.py (hunk 5 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py]: Unfortunately, we need this for forward passes through the parent modules, which are not always Linear layers. So let's leave as is for now, and we can tackle this with more hooks to inputs to each of the smoothing layers in a follow-up?

File: src/llmcompressor/modifiers/awq/base.py [hunk 10]
```diff
@@ -49,32 +53,50 @@ class AWQModifier(Modifier):
...
+                - raise error if any unused activations remain
+        - on_end
+            - re-run logic of sequential epoch end (in case of basic pipeline)
+            - set scales and zero points
+            - remove activation hooks
         - on_finalize
             - clear resolved mappings and captured activations
 
+    :param sequential_targets: list of layer names to compress during AWQ, or
+        '__ALL__' to compress every layer in the model
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:99]:
```suggestion
    :param sequential_targets: list of module names to compress in the same calibration pass
```

There might be a better way to describe this in english

File: src/llmcompressor/modifiers/awq/base.py [hunk 11]
```diff
@@ -96,53 +118,147 @@ class AWQModifier(Modifier):
...
 
+        # apply config to model and prepare calibration hooks
+        if QuantizationMixin.has_config(self):
+            QuantizationMixin.initialize_quantization(self, state.model)
+
         self._set_resolved_mappings(state.model)
 
-        with calibration_forward_context(state.model):
-            self._set_module_kwargs(state.model, state.data.calib)
+        self._set_module_kwargs(state.model, state.data.calib)
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/awq/base.py:199]: This `_set_module_kwargs` function still needs work. This code, as currently written, still runs through the entire dataset (with massive batch size). This means that large number of calibration samples will still trigger oom.

**Review by kylesayrs — CHANGES_REQUESTED** [review]

File: src/llmcompressor/modifiers/awq/base.py (hunk 11 above)
Comment (kylesayrs) [reply src/llmcompressor/modifiers/awq/base.py:199]: It aught to be possible to capture these kwargs during the calibration step

File: src/llmcompressor/modifiers/awq/base.py (hunk 10 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:99]: Will use your suggestion, thanks! I pulled it directly from [GPTQ](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modifiers/quantization/gptq/base.py#L68). Should I change it there as well?

File: src/llmcompressor/modifiers/awq/base.py (hunk 11 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:198]: It only does it for the first module, and errors out after kwargs are cached. certainly not the greatest, and we can improve, but I'd like to do that in a follow-up

**Review by kylesayrs — APPROVED** [review]
This modifier still has extreme memory constraints due to how we're still megabatching everything for `_set_module_kwargs` and `_apply_smoothing`

This followup PR gets part way towards removing that, but still batches outputs in its current state (grep for `torch.cat`): https://github.com/vllm-project/llm-compressor/pull/1435

Let's land this as is and add accumulation in a follow up to reduce memory requirements

File: src/llmcompressor/modifiers/awq/base.py (hunk 11 above)
Comment (brian-dellabetta) [reply src/llmcompressor/modifiers/awq/base.py:217]: We discussed that we'll address this in https://github.com/vllm-project/llm-compressor/pull/1435, so i'm going to resolve this comment

**Review by rahul-tuli — APPROVED** [review]
Lgtm! Good updates!
